Analyzer Plugins

lrytz opened the pull request:
Resubmission of #2005, incorporated the points discussed:
- `debuglog` and more documentation about function types with refinements in UnCurry
  - new: removed an unnecessary `if` in UnCurry
- `AnnotationCheckers` are source compatible with the way they were before. I checked that this works by running the un-modified version of the continuations plugin.
- Removed the hook into `SymbolTable.phase_=`
- Word-Wrapped commit comments
- Fixed Typos

lrytz committed:
replace symbols correctly when subtyping dependent types

as discussed in [1], this change has no impact on the scala type
system, but it can fix subtyping for annotation checkers if the
pluggable type system supports annotations that can refer to method
parameters.

[1] https://groups.google.com/forum/#!topic/scala-internals/kSJLzYkmif0/discussion

lrytz committed:
case module toString is synthetic

lrytz committed:
Allow for Function treess with refined types in UnCurry.

Also removes an unnecessary condition in UnCurry, isFunctionType(fun.tpe)
is always true.

lrytz committed:
Keep annotations when computing lubs

Integrates annotationsLub into lub.

Also fixes SubstSymMap when mapping over annotaion trees. I don't
understand what the previous code was supposed to achieve, but it
crashed in some of my examples.

lrytz committed:
SI-1803, plus documentation and cleanups in Namers, mainly in typeSig

- when typing (and naming) a ValDef, tpt and rhs are now type checked
in the same context (the inner / ValDef context). this does not change
any behavior, but is more uniform (same as for DefDef). martin told me
(offline) that this change is desirable if it doesn't break anything.
(it doesn't).

- typeSig is now more uniform with a separate method for each case
(methodSig, valDefSig, etc). methodSig was cleaned up (no more variables)
and documented. the type returned by methodSig no longer contains /
refers to type skolems, but to the actual type parameters (so we don't
need to replace the skolems lateron).

- documentation on constructor contexts, type skolems

- more tests for SI-5543

lrytz committed:
Analyzer Plugins

AnnotationCheckers are insufficient because they live outside the
compiler cake and it's not possible to pass a Typer into an annotation
checker.

Analyzer plugins hook into important places of the compiler:

  - when the namer assigns a type to a symbol (plus a special hook for
    accessors)
  - before typing a tree, to modify the expected type
  - after typing a tree, to modify the type assigned to the tree

Analyzer plugins and annotation checker can be activated only during
selected phases of the compiler.

Refactored the CPS plugin to use an analyzer plugin (since
adaptToAnnotations is now part of analyzer plugins, no longer
annotation checkers).

lrytz commented:
review by @paulp, @retronym, @xeno-by.

paulp commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,216 @@
+     * @param mode  Mode that was used for typing `tree`
+     * @param pt    Expected type that was used for typing `tree`
+     */
+    def pluginsTyped(tpe: Type, typer: Typer, tree: Tree, mode: Int, pt: Type): Type = tpe
```
I guess it's so darn topical, I have to ask. Can you look at my two candidates for being "analyzer plugins"? I speak of course of interactive and scaladoc. This is not proud code, but it does encapsulate all the little bits which used to be scattered hither and thither with "forScaladoc" and "forInteractive" guards.

My question is, would either of these be able to use the "AnalyzerPlugins" framework to do what they do?

```
trait InteractiveAnalyzer extends Analyzer {
  val global : Global
  import global._

  override def newTyper(context: Context): InteractiveTyper = new InteractiveTyper(context)
  override def newNamer(context: Context): InteractiveNamer = new InteractiveNamer(context)
  override protected def newPatternMatching = false

  class InteractiveTyper(context0: Context) extends Typer(context0) {
    override def canAdaptConstantTypeToLiteral = false
    override def canTranslateEmptyListToNil    = false
    override def missingSelectErrorTree(tree: Tree, qual: Tree, name: Name): Tree = tree match {
      case Select(_, _)             => treeCopy.Select(tree, qual, name)
      case SelectFromTypeTree(_, _) => treeCopy.SelectFromTypeTree(tree, qual, name)
    }
  }

  class InteractiveNamer(context0: Context) extends Namer(context0) {
    override def saveDefaultGetter(meth: Symbol, default: Symbol) {
      // save the default getters as attachments in the method symbol. if compiling the
      // same local block several times (which can happen in interactive mode) we might
      // otherwise not find the default symbol, because the second time it the method
      // symbol will be re-entered in the scope but the default parameter will not.
      meth.attachments.get[DefaultsOfLocalMethodAttachment] match {
        case Some(att) => att.defaultGetters += default
        case None      => meth.updateAttachment(new DefaultsOfLocalMethodAttachment(default))
      }
    }
    // this logic is needed in case typer was interrupted half
    // way through and then comes back to do the tree again. In
    // that case the definitions that were already attributed as
    // well as any default parameters of such methods need to be
    // re-entered in the current scope.
    override def enterExistingSym(sym: Symbol): Context = {
      if (sym != null && sym.owner.isTerm) {
        enterIfNotThere(sym)
        if (sym.isLazy)
          sym.lazyAccessor andAlso enterIfNotThere

        for (defAtt <- sym.attachments.get[DefaultsOfLocalMethodAttachment])
          defAtt.defaultGetters foreach enterIfNotThere
      }
      super.enterExistingSym(sym)
    }
    override def enterIfNotThere(sym: Symbol) {
      val scope = context.scope
      @tailrec def search(e: ScopeEntry) {
        if ((e eq null) || (e.owner ne scope))
          scope enter sym
        else if (e.sym ne sym)  // otherwise, aborts since we found sym
          search(e.tail)
      }
      search(scope lookupEntry sym.name)
    }
  }
}
```

```
trait ScaladocAnalyzer extends Analyzer {
  val global : ScaladocGlobal
  import global._

  override def newTyper(context: Context): ScaladocTyper = new ScaladocTyper(context)

  class ScaladocTyper(context0: Context) extends Typer(context0) {
    private def unit = context.unit

    override def canAdaptConstantTypeToLiteral = false

    override def typedDocDef(docDef: DocDef, mode: Mode, pt: Type): Tree = {
      val sym = docDef.symbol

      if ((sym ne null) && (sym ne NoSymbol)) {
        val comment = docDef.comment
        docComments(sym) = comment
        comment.defineVariables(sym)
        val typer1 = newTyper(context.makeNewScope(docDef, context.owner))
        for (useCase <- comment.useCases) {
          typer1.silent(_ => typer1 defineUseCases useCase) match {
            case SilentTypeError(err) =>
              unit.warning(useCase.pos, err.errMsg)
            case _ =>
          }
          for (useCaseSym <- useCase.defined) {
            if (sym.name != useCaseSym.name)
              unit.warning(useCase.pos, "@usecase " + useCaseSym.name.decode + " does not match commented symbol: " + sym.name.decode)
          }
        }
      }

      super.typedDocDef(docDef, mode, pt)
    }

    def defineUseCases(useCase: UseCase): List[Symbol] = {
      def stringParser(str: String): syntaxAnalyzer.Parser = {
        val file = new BatchSourceFile(context.unit.source.file, str) {
          override def positionInUltimateSource(pos: Position) = {
            pos.withSource(context.unit.source, useCase.pos.start)
          }
        }
        newUnitParser(new CompilationUnit(file))
      }

      val trees = stringParser(useCase.body+";").nonLocalDefOrDcl
      val enclClass = context.enclClass.owner

      def defineAlias(name: Name) = (
        if (context.scope.lookup(name) == NoSymbol) {
          lookupVariable(name.toString.substring(1), enclClass) foreach { repl =>
            silent(_.typedTypeConstructor(stringParser(repl).typ())) map { tpt =>
              val alias = enclClass.newAliasType(name.toTypeName, useCase.pos)
              val tparams = cloneSymbolsAtOwner(tpt.tpe.typeSymbol.typeParams, alias)
              val newInfo = genPolyType(tparams, appliedType(tpt.tpe, tparams map (_.tpe)))
              alias setInfo newInfo
              context.scope.enter(alias)
            }
          }
        }
      )

      for (tree <- trees; t <- tree)
        t match {
          case Ident(name) if name startsWith '$' => defineAlias(name)
          case _ =>
        }

      useCase.aliases = context.scope.toList
      namer.enterSyms(trees)
      typedStats(trees, NoSymbol)
      useCase.defined = context.scope.toList filterNot (useCase.aliases contains _)

      if (settings.debug.value)
        useCase.defined foreach (sym => println("defined use cases: %s:%s".format(sym, sym.tpe)))

      useCase.defined
    }
  }
}
```

lrytz replied:
The main difference is that I need more fine-grained hooks; it could be done but would require refactorings in Namers/Typers which go way beyond what I did now (see e.g. invocations of `pluginsTypeSig` or `pluginsTyped`).

But yes, things go into a similar direction. Another axis is macro annotations where we want to be able to do similar things (change types for symbols) - I'll talk to Eugene about this. I'm not worried because analyzer plugins are quite well delimited, and only exploring the different requirements will help us find convergence eventually.

paulp commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,216 @@
+     * mode (see method adapt in trait Typers).
+     *
+     * An implementation cannot rely on canAdaptAnnotations being called before. If the implementing
+     * class cannot do the adaptiong, it should return the tree unchanged.
```
adapting

paulp commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,216 @@
+
+    /**
+     * Modify the type of a return expression. By default, return expressins have type
+     * NothingClass.tpe.
```
expressions

paulp commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,216 @@
+     *
+     * @param defTree is the definition for which the type was computed. The different cases are
+     * outlined below. Note that this type is untyped (for methods and values with inferred type,
+     * the typed rhs trees are availalble in analyzer.transformed).
```
available

paulp commented on src/compiler/scala/tools/nsc/typechecker/Namers.scala:
```diff
@@ -1213,9 +1333,32 @@ trait Namers extends MethodSynthesis {
+      pluginsTypeSig(res, typer, tdef, WildcardType)
     }
 
+    private def importSig(iprt: Import) = {
```
I suppose this goes in the pedantry column, but I strongly dislike semi-random letter strings like "iprt". Most things have a standard name or two for locals; in this case, I like "imp".

lrytz replied:
sure, i don't particularly like iprt either

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2474/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1738/

paulp commented on src/reflect/scala/reflect/internal/Types.scala:
```diff
@@ -6613,10 +6603,20 @@ trait Types extends api.Types { self: SymbolTable =>
+        else AnnotatedType(anns, tp1, selfs)
+      case _ =>
+        elimAnonymousClass(t.underlying)
+    }
```
Can you elaborate on the motivation for attempting to eliminate those things specifically?

This for instance doesn't work, but maybe it doesn't matter:

```
scala> final val x = "abc"
x: String("abc") = abc

scala> elimSingletonAndAnonClass(typeOf[x.type])
res8: $r.intp.global.Type = String

scala> type Foo = x.type
defined type alias Foo

scala> elimSingletonAndAnonClass(typeOf[Foo])
res9: $r.intp.global.Type = Foo

scala> res9.dealias
res10: $r.intp.global.Type = String("abc")
```

I always have that one on the brain due to dealiasWiden.

lrytz replied:
I actually don't know. The only change introduced by here is that type annotations are not eliminated.

Actually I probably got it wrong in my comment: "eliminates singleton types using _.underlying". Underlying means all kinds of different things (for ThisType, AnnotatedType, ConstantType, TypeProxy, ...) and I'm quite sure now that the person who wrote the code should have picked something else than `underlying` here. But I'm not sure what was the intention..

lrytz replied:
I guess the intention was indeed to `widen` singleton types: https://github.com/scala/scala/commit/7529035#L0L3133

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1741/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2477/

paulp commented:
I feel like an addition of this magnitude warrants a more extensive test to exercise it and demonstrate what uses are in the mind of the author.

On the whole it looks quite good though. I was fairly nervous going in, now I'm so relaxed I might fall asleep in my chair. So this has my provisional lgtm assuming nothing too shocking takes place from here.

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1738/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1741/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2474/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2477/

lrytz commented:
now using `dealiasWiden` instead of `underlying` in `lub/elimSub`. fixed typos.

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1768/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1768/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2514/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2514/

lrytz commented:
seems multiple-personality kittie is happy.

lrytz committed:
Fix context for type checking early initializers

Before:

scala> class A { class C extends { val x: A = this } with AnyRef }
<console>:7: error: type mismatch;
 found   : A.this.C
 required: A
       class A { class C extends { val x: A = this } with AnyRef }
                                              ^

Note that the same thing is necessary and already done in Namers (see
def createNamer). The whole logic of when and how to create contexts
should be factored out and reused in Namer and Typer.

(
My Hobby [1]: detecting compiler by just looking at its soruce
[1] http://www.explainxkcd.com/wiki/index.php?title=Category:My_Hobby
)

lrytz commented:
added a fix for another bug i found. final review by @retronym ?

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1826/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2562/

scala-jenkins commented:
jenkins job pr-rangepos: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1826/

scala-jenkins commented:
jenkins job pr-scala-testsuite-linux-opt: Success - https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2562/

retronym commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,217 @@
+    val annotCheckersType = adaptTypeOfReturn(tree.expr, pt, tpe)
+    analyzerPlugins.foldLeft(annotCheckersType)((tpe, plugin) =>
+      if (!plugin.isActive()) tpe else plugin.pluginsTypedReturn(tpe, typer, tree, pt))
+  }
```
It's hard to know without profiling, but perhaps we should treat the common case (ie no analyzer plugins) specially to avoid the call to `foldLeft` and the closure allocation.

lrytz replied:
True. I'll introduce a fast path for all those methods in AnalyzerPlugins and AnnotationCheckers.

retronym commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,217 @@
+     * assigning `tpe` to `defTree.symbol` (untested) - the final type computed by this method
+     * will then be assigned to the definition's symbol by monoTypeCompleter (in Namers).
+     *
+     * The hooks into typeSig` allow analyzer plugins to add annotations to (or change the types
```
You're missing a backtick here. You've done a good job writing these docs, you should make sure they render nicely, too. You might add a few links and a few more backticks along the way.

lrytz replied:
OK, I didn't actually run scaladoc. I'll do that and adjust the comment syntax.

retronym commented:
Aside from those small comments, lgtm.

lrytz commented:
Thanks for the review. I'll do the discussed changes, wait for kitty to do its job and then merge. After that I'll prepare a PR for master that includes the same commits.

paulp commented on src/compiler/scala/tools/nsc/typechecker/AnalyzerPlugins.scala:
```diff
@@ -0,0 +1,217 @@
+  import global._
+
+
+  abstract class AnalyzerPlugin {
```
I didn't notice this before. Can this be a trait? Without performance data to say otherwise, trait should be preferred.

lrytz replied:
Sure, it (and `AnnotationCheckers`) can be traits.

paulp commented:
Also, I was still sort of hoping to see some test coverage of the AnalyzerPlugin trait/class. Having a working analyzer plugin which exercises each method in the interface should be IMO mandatory before merging.

lrytz commented:
I agree we should have tests, I just see right now how to write them.. I'll spend some time looking into it. If somebody has ideas or knows tools that could help, please send me some pointers; I'm not really familiar with the more advanced ways for testing the compiler (i.e. the tests where you have a compiler instance at hand).

lrytz commented:
ok, i think i got it how to write a test.

lrytz commented:
- Added a test for analyzer plugins and annotation checkers
- Fast-track the hooks in the common case when there are no registered plugins
- Fixed typo and verified that scaladoc comments render

lrytz commented:
rebased on top of the current `git merge-base master 2.10.x` - this will make things easier when merging into master.

lrytz commented:
PLS REBUILD ALL, :cat2:

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1863/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2601/

scala-jenkins commented:
Started jenkins job pr-rangepos at https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1865/

scala-jenkins commented:
Started jenkins job pr-scala-testsuite-linux-opt at https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2603/

scala-jenkins commented:
jenkins job pr-rangepos: ABORTED - https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1865/
![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)
SPURIOUS ABORT? -- PLS REBUILD pr-rangepos

lrytz commented:
kitty is a bit confused, but both jobs succeeded (pr-rangepos/1863 and pr-scala-testsuite-linux-opt/2601). So it's ready to merge.

scala-jenkins commented:
(kitty-note-to-self: ignore 13054225)
:cat: Roger! Rebuilding pr-rangepos-per-commit, pr-checkin-per-commit. :rotating_light:

scala-jenkins commented:
(kitty-note-to-self: ignore 13056421)
:cat: Roger! Rebuilding . :rotating_light:

scala-jenkins commented:
Job pr-checkin-per-commit failed for a06d31:
Failed tests:
- presentation/implicit-member
- presentation/callcc-interpreter
  Took 176 min.<br> ![sad kitty](http://cdn.memegenerator.net/instances/100x/31464013.jpg)

xeno-by commented:
gz :)

lrytz commented:
gz?

xeno-by commented:
short for "congratulations"! :+1:

adriaanm commented:
i think the kitty was right for the commit it mentioned:

https://scala-webapps.epfl.ch/jenkins/view/pr-validators/job/pr-checkin-per-commit/29/consoleText

lrytz commented:
both builds finished successfully (https://scala-webapps.epfl.ch/jenkins/job/pr-rangepos/1863 and https://scala-webapps.epfl.ch/jenkins/job/pr-scala-testsuite-linux-opt/2601)

also the failures have some `TIMEOUT` message, so i thing they're spurious

adriaanm commented:
the build kitten is now building individual commits, the job i link to was building commit a06d31

```
+ git checkout -B pull-2022-merge
+ git fetch origin +pull/2022/head:pullreq/2022
+ git log --oneline -n 1 a06d31f6a2df021acc18b10ad9cad0f80d7498f2
a06d31f Keep annotations when computing lubs
+ git merge a06d31f6a2df021acc18b10ad9cad0f80d7498f2
```

the failures:

```
  [partest] testing: [...]/files/presentation/implicit-member                     [FAILED]
  [partest] reload: ImplicitMember.scala
  [partest] 
  [partest] askTypeCompletion at ImplicitMember.scala(7,7)
  [partest] TIMEOUT: ImplicitMember.scala:7:7
  [partest] testing: [...]/files/presentation/callcc-interpreter                  [FAILED]
  [partest] reload: CallccInterpreter.scala
  [partest] 
  [partest] askTypeCompletion at CallccInterpreter.scala(51,38)
  [partest] TIMEOUT: CallccInterpreter.scala:51:38
  [partest] 
  [partest] askType at CallccInterpreter.scala(14,21)
  [partest] ================================================================================
  [partest] [response] askTypeAt at (14,21)
  [partest] def unitM[A >: Nothing <: Any](a: A): callccInterpreter.M[A] = callccInterpreter.this.M.apply[A](((c: A => callccInterpreter.Answer) => c.apply(a)))
  [partest] ================================================================================
  [partest] 
  [partest] askType at CallccInterpreter.scala(16,12)
  [partest] ================================================================================
  [partest] [response] askTypeAt at (16,12)
  [partest] tree not found at source-/localhome/jenkins/b/workspace/pr-checkin-per-commit/test/files/presentation/callcc-interpreter/src/CallccInterpreter.scala,line-51,offset=1451
  [partest] def id[A >: Nothing <: Any]: A => A = ((x: A) => x)
  [partest] ================================================================================
  [partest] 
  [partest] askType at CallccInterpreter.scala(17,25)
  [partest] ================================================================================
  [partest] [response] askTypeAt at (17,25)
  [partest] def showM(m: M[Value]) = m.in(id).toString()
  [partest] ================================================================================
  [partest] 
  [partest] askType at CallccInterpreter.scala(50,30)
  [partest] ================================================================================
  [partest] [response] askTypeAt at (50,30)
  [partest] def add(a: callccInterpreter.Value, b: callccInterpreter.Value): callccInterpreter.M[_ >: callccInterpreter.Num with callccInterpreter.Wrong.type <: Product with Serializable with callccInterpreter.Value] = scala.this.Predef.Pair.apply[callccInterpreter.Value, callccInterpreter.Value](a, b) match {
  [partest]   case scala.this.Predef.Pair.unapply[callccInterpreter.Value, callccInterpreter.Value](<unapply-selector>) <unapply> ((n: Int)callccInterpreter.Num((m @ _)), (n: Int)callccInterpreter.Num((n @ _))) => this.unitM[callccInterpreter.Num](callccInterpreter.this.Num.apply(m.+(n)))
  [partest]   case _ => callccInterpreter.this.unitM[callccInterpreter.Wrong.type](callccInterpreter.this.Wrong)
  [partest] }
  [partest] ================================================================================
```

lrytz commented:
right, i saw that

adriaanm commented:
maybe we should get rid of those tests then /cc @dragos

adriaanm commented:
(or make them more robust)

lrytz commented:
did you see them fail often? i always skim through test failures here and there, and these ones are rare, I think.

xeno-by commented:
they happened today in my pull request

lrytz commented:
can you give me a link to that job? it seems all those failures happen on builds on the "chara" slave (also here: https://github.com/scala/scala/pull/2017), so that machine can't handle 4 concurrent builds.

xeno-by commented:
https://github.com/scala/scala/pull/2054

lrytz commented:
yep, chara as well.

dragos commented:
The timeout is by default 10 seconds. I wouldn't remove those tests, they are pretty basic. We can increase the timeout, if the machine is slow.
